File: Problems 0200-0299/224. Basic Calculator/solution.py
class Solution:
    def calculate(self, s: str) -> int:
        stack = []
        # running result in current "level"/context
        res = 0
        curr = 0
        sign = 1
        for c in s:
            if c.isnumeric():
                curr = curr*10 + int(c)
            elif c == "+" or c == "-":
                res += curr * sign
                # Reset
                curr = 0
                sign = 1 if c == "+" else -1
            elif c == "(":
                # Start new "context"
                stack.append(res)
                stack.append(sign)
                # Reset
                sign = 1
                # Reset res as we are in a new context/level
                res = 0
            elif c == ")":
                res += sign * curr
                curr = 0
                res *= stack.pop()
                res += stack.pop()

        if curr != 0:
            res += sign * curr
        return res

File: Problems 0200-0299/224. Basic Calculator/solution.cpp
class Solution {
    public:
        int calculate(string s) {
            stack<int> stack;
            // Running sum at current level/context: () -> is a new level/context
            int res = 0;
            int num = 0;
            int sign = 1;
            for (char c : s) {
                if (isdigit(c)) {
                    num = num*10 + (c-'0');
                } else if (c == '+' || c == '-') {
                    res += num*sign;
                    sign = c == '+' ? 1 : -1;
                    num = 0;
                } else if (c == '(') {
                    // start new context
                    stack.push(res);
                    stack.push(sign);
                    sign = 1;
                    res = 0;
                } else if (c == ')') {
                    res += num*sign;
                    num = 0;
                    res = stack.top()*res;
                    stack.pop();
                    res += stack.top();
                    stack.pop();
                }
            }
    
            if (num != 0) res += num * sign;
            return res;
        }
    };